core/src/main/scala/org/ensime/core/Indexer.scala

```diff
@@ -24,10 +24,26 @@ class Indexer(
     LineSourcePositionHelper.fromFqnSymbol(hit)(vfs)
   )
 
-  def oldSearchTypes(query: String, max: Int) =
-    index.searchClasses(query, max).filterNot {
-      name => name.fqn.endsWith("$") || name.fqn.endsWith("$class")
-    }.map(typeResult)
+  def oldSearchTypes(query: String, max: Int) = {
+    val scalaNames = scala.collection.mutable.Set.empty[String]
+
+    def isUnique(fqn: String): Boolean = {
+      val stripped =
+        if (fqn.endsWith("$")) fqn.dropRight(1)
+        else if (fqn.endsWith("$class")) fqn.dropRight(6)
+        else fqn
+
+      if (scalaNames.contains(stripped)) {
+        false
+      } else {
+        scalaNames.add(stripped)
+        true
+      }
+    }
+
+    index.searchClasses(query, max)
```

Comment:
yeah, I'd prefer this... I don't think performance is too much of a concern here.  But instead of a Set (which reorders things) maybe just `.distinct` ?